File: orig_app/internal/handler/webhooks/hook_log_actions.go
package webhooks

import (
	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
	"gitlab.com/gitlab-org/gitlab-redux/internal/util"
)

// HookLogActions handles actions related to webhook logs
type HookLogActions struct {
	resendService *service.WebHooksEventsResendService
	executionNotice *HookExecutionNotice
	logger *util.Logger
}

// NewHookLogActions creates a new instance of HookLogActions
func NewHookLogActions(resendService *service.WebHooksEventsResendService, executionNotice *HookExecutionNotice, logger *util.Logger) *HookLogActions {
	return &HookLogActions{
		resendService: resendService,
		executionNotice: executionNotice,
		logger: logger,
	}
}

// Show displays a webhook log
func (h *HookLogActions) Show(ctx *gin.Context, hook model.WebHook) {
	// Hide search settings
	ctx.Set("hide_search_settings", true)

	// Get the hook log ID from the URL parameters
	hookLogID := ctx.Param("id")

	// Find the hook log
	hookLog, err := h.getHookLog(hook, hookLogID)
	if err != nil {
		ctx.JSON(404, gin.H{"error": "Hook log not found"})
		return
	}

	// Render the show template
	ctx.HTML(200, "webhooks/logs/show.html", gin.H{
		"hook": hook,
		"hookLog": hookLog,
		"hide_search_settings": true,
	})
}

// Retry retries a webhook execution
func (h *HookLogActions) Retry(ctx *gin.Context, hook model.WebHook) {
	// Get the hook log ID from the URL parameters
	hookLogID := ctx.Param("id")

	// Find the hook log
	hookLog, err := h.getHookLog(hook, hookLogID)
	if err != nil {
		ctx.JSON(404, gin.H{"error": "Hook log not found"})
		return
	}

	// Get the current user from the context
	currentUser, _ := ctx.Get("current_user")
	user, _ := currentUser.(model.User)

	// Execute the hook
	result := h.executeHook(ctx, hookLog, user)

	if result.Success {
		// Redirect to the after retry path
		ctx.Redirect(302, h.getAfterRetryRedirectPath(ctx, hook))
	} else {
		// Set a warning flash message
		ctx.Set("flash_warning", result.Message)

		// Redirect back with a fallback
		ctx.Redirect(302, h.getAfterRetryRedirectPath(ctx, hook))
	}
}

// getHookLog retrieves a webhook log by ID
func (h *HookLogActions) getHookLog(hook model.WebHook, hookLogID string) (model.WebHookLog, error) {
	return hook.GetWebHookLogs().Find(hookLogID)
}

// executeHook executes a webhook and sets the execution notice
func (h *HookLogActions) executeHook(ctx *gin.Context, hookLog model.WebHookLog, user model.User) service.ServiceResponse {
	// Execute the hook
	result := h.resendService.Execute(hookLog, user)

	// Set the hook execution notice
	h.executionNotice.SetHookExecutionNotice(ctx, result.ToMap())

	return result
}

// getAfterRetryRedirectPath returns the path to redirect to after a retry
func (h *HookLogActions) getAfterRetryRedirectPath(ctx *gin.Context, hook model.WebHook) string {
	// This would typically be a method on the hook model or a helper function
	// For now, we'll return a default path
	return "/webhooks/" + hook.ID() + "/logs"
}

File: rust_app/src/services/webhooks/hook_log_actions.rs
// Ported from orig_app/app/controllers/concerns/web_hooks/hook_log_actions.rb
// Provides actions for showing and retrying webhook logs in a controller/service context.
// Ported: 2025-04-24

use super::hook_execution_notice::HookExecutionNotice;
use crate::models::user::User;
use crate::webhooks::{WebHook, WebHookLog, WebHookResult};
use actix_web::{web, HttpResponse};
use serde_json::json;

pub trait HookLogActions: HookExecutionNotice {
    // Returns the current WebHook instance
    fn hook(&self) -> &WebHook;

    // Show a webhook log (hides search settings)
    fn show(&self, log_id: i32) -> HttpResponse {
        // In a real implementation, hide_search_settings would set a flag in the context.
        HttpResponse::Ok().json(json!({
            "hook_log": self.get_hook_log(log_id)
        }))
    }

    // Retry a webhook log execution
    fn retry(&self, log_id: i32, user: &User) -> HttpResponse {
        let result = self.execute_hook(log_id, user);
        if result.success {
            HttpResponse::Found()
                .header("Location", self.after_retry_redirect_path())
                .finish()
        } else {
            HttpResponse::Found()
                .header("Location", self.after_retry_redirect_path())
                .json(json!({
                    "warning": result.message
                }))
        }
    }

    // Retrieve a webhook log by ID
    fn get_hook_log(&self, log_id: i32) -> &WebHookLog {
        // TODO: Implement hook log retrieval from database
        unimplemented!()
    }

    // Execute the webhook log (resend service)
    fn execute_hook(&self, log_id: i32, user: &User) -> WebHookResult {
        // TODO: Implement WebHooks::Events::ResendService logic
        let result = WebHookResult {
            success: true,
            message: "Hook executed".to_string(),
            http_status: Some(200),
        };
        self.set_hook_execution_notice(&result);
        result
    }

    // Path to redirect to after retry
    fn after_retry_redirect_path(&self) -> String {
        // TODO: Implement based on your routing needs
        "/webhooks".to_string()
    }
}
// End of port from orig_app/app/controllers/concerns/web_hooks/hook_log_actions.rb
